src/arithmetic/algebraic_expression.c

```diff
@@ -310,93 +209,301 @@ void AlgebraicExpression_PrependTerm(AlgebraicExpression *ae, GrB_Matrix m, bool
+            expressions = array_append(expressions, iexp);
+            expIdx++;
         }
+    }
+
+    return expressions;
+}
 
-        for(int i = 0; i < hops; i++) {
-            bool freeMatrix = (labelCount > 1);
-            AlgebraicExpression_AppendTerm(exp, mat, transpose, freeMatrix);
+static AlgebraicExpressionOperand _AlgebraicExpression_OperandFromNode(Node *n) {
+    AlgebraicExpressionOperand op;   
+    op.free = false;
+    op.transpose = false;
+    op.operand = Node_GetMatrix(n);
+    return op;
+}
+
+static AlgebraicExpressionOperand _AlgebraicExpression_OperandFromEdge(
+    Edge *e,
+    bool transpose,
+    const AST* ast
+) {
+    AlgebraicExpressionOperand op;
+
+    // TODO: find a better way for retriving AST graph entity.
+    GrB_Matrix mat = Edge_GetMatrix(e);
+    AST_LinkEntity *astEdge = (AST_LinkEntity*)MatchClause_GetEntity(ast->matchNode, e->alias);
+    assert(astEdge);
+
+    int labelCount = AST_LinkEntity_LabelCount(astEdge);
+    bool freeMatrix = (labelCount > 1);
+    // [:A|:B]
+    // Create matrix M = A+B.
+    if(labelCount > 1) {
+        GraphContext *gc = GraphContext_GetFromTLS();
+        Graph *g = gc->g;
+
+        GrB_Matrix m;
+        GrB_Matrix_new(&m, GrB_BOOL, Graph_RequiredMatrixDim(g), Graph_RequiredMatrixDim(g));
+
+        for(int i = 0; i < labelCount; i++) {
+            char *label = astEdge->labels[i];
+            Schema *s = GraphContext_GetSchema(gc, label, SCHEMA_EDGE);
+            if(!s) continue;
+            GrB_Matrix l = Graph_GetRelationMatrix(g, s->id);
+            GrB_Info info = GrB_eWiseAdd_Matrix_Semiring(m, NULL, NULL, Rg_structured_bool, m, l, NULL);
         }
+        mat = m;
+    }
+
+    op.operand = mat;
+    op.free = freeMatrix;
+    op.transpose = transpose;
+    return op;
+}
+
+AlgebraicExpression *AlgebraicExpression_Empty() {
+    return _AE_MUL(1);
+}
+
+AlgebraicExpression **AlgebraicExpression_From_QueryGraph(const QueryGraph *qg, const AST *ast, size_t *exp_count) {
+    /* Construct algebraic expression(s) from query graph 
+     * trying to take advantage of long multiplications with as few 
+     * transpose as possible we'll transform paths crossing the graph 
+     * "diameter", these are guarantee to be the longest, although
+     * there might be situations in which these are not the most optimal paths
+     * to explore.
+     * 
+     * Once a path been transformed it's removed from the graph and the process
+     * repeat itself. */
+
+    assert(qg);
+    
+    /* A graph with no edges implies an empty algebraic expression
+     * the reasoning behind this decission is that algebraic expression 
+     * represent graph traversals, no edges means no traversals. */
+    if(qg->edge_count == 0) {
+        *exp_count = 0;
+        return NULL;
+    }
 
-        if(dest->label) {
-            GrB_Matrix destMat = Node_GetMatrix(dest);
-            AlgebraicExpression_AppendTerm(exp, destMat, false, false);
+    Edge *e;
```

Comment:
Please NULL-initialize, as this causes a `may be used uninitialized` compiler warning for me (unnecessarily, as `level` will always be > 0).